src/service/resource.js

```diff
@@ -654,51 +654,60 @@ export class Resource {
   }
 
   /**
+   * @param {number|boolean} viewport derived from renderOutsideViewport.
    * @return {!Promise} resolves when underlying element is built and within the
-   *    range specified by implementation's renderOutsideViewport
+   *    viewport range given.
    */
-  whenWithinRenderOutsideViewport() {
-    if (!this.isLayoutPending()) {
+  whenWithinViewport(viewport) {
+    dev().assert(viewport !== false);
+    // Resolve is already laid out or viewport is true.
+    if (!this.isLayoutPending() || viewport === true) {
       return Promise.resolve();
     }
-    if (this.renderOutsideViewportPromise_) {
-      return this.renderOutsideViewportPromise_;
+    // See if pre-existing promise.
+    const viewportNum = dev().assertNumber(viewport);
+    if (this.withViewportDeferreds_[viewportNum]) {
+      return this.withViewportDeferreds_[viewportNum].promise;
     }
-    const deferred = new Deferred();
-    this.renderOutsideViewportResolve_ = deferred.resolve;
-    return this.renderOutsideViewportPromise_ = deferred.promise;
+    // See if already within viewport multiplier.
+    if (this.withinViewportMultipliers([viewport])[0]) {
```

Comment:
I don't think this is a good approach. `withinViewportMultipliers` is called a lot, and this will introduce a whole lot of GC pressure and needless loops. It seems like we should change `withinViewportMultipliers` to a `distanceFromViewport`, and do the calculations in the `resolveDeferredsWhenWithinViewports_`.